RevEng: Support Type Aliases

lajones opened the pull request:
Fix for #3843 - allow columns to use user-defined data types as the column type.

Note: I'm open to reducing this down to only the `SqlServerTypeMapper`, `SqlServerDatabaseModelFactory` etc. But I wanted to put this wider approach out there so that other providers could re-use the same architecture i.e. they would only have to provide their own implementation of `XyzDatabaseModelFactory.GetUserDefinedDataTypes()` and after that it would "just work". The price is `GetUserDefinedDataTypes()` has to run before the first call to `Create()` and so it has to manage its own connection.

lajones commented:
Note to provider-writers - with this approach you will need to implement ` XyzDatabaseModelFactory.GetUserDefinedDataTypes(string)` - it's safe to just return null if you have no such types.

roji commented:
Here's an Npgsql perspective... PostgreSQL supports user-defined types beyond simple aliases (composite types, enums). Since these vary from database to database, it would be good if there were a way to load them as well (with a connection string). In other words, it's oddly specific to have a mechanism so specific for loading user-defined data types with a connection string while neglecting all other cases.

Also, the name "user-defined data types" has no meaning outside SqlServer (and seems to imply something much more than a simple alias). Perhaps "type aliases" is a more generic naming?

Finally, I may be missing stuff, but design-wise isn't it simpler/better to have the database model factory return type aliases as part the DatabaseModel returned by the Create method, rather than defining an additional separate GetUserDefinedDataTypes?

ilmax commented:
@roji if I understand the topic correctly, [Oracle](https://docs.oracle.com/cd/A91202_01/901_doc/server.901/a88856/c14ordb.htm) use the same name for the same thing, so (at least to me) the name UDT looks good.

roji commented:
@ilmax be careful, the EF7 is discussed user-defined *data* types here - these are simple aliases. User-defined types (without data) is a different SqlServer feature that is not supported. See https://github.com/aspnet/EntityFramework/issues/3843#issuecomment-161086035.

lajones commented:
@ilmax @roji is correct. These are just aliases for underlying system types - we are not supporting user-defined types just yet. And thanks @roji - I like the name `TypeAlias`.

Let me look into the idea of just adding `TypeAliases` to the `DatabaseModel` instead and I'll post a response here.

ilmax commented:
@roji thank you for the clarification, so you are right, a better name should help avoid confusion!

roji commented:
A development on something I said above...

Type aliases means that you support database-specific things in the TypeMapper - types that are present in one database but not in another. If you allow this, you may want to consider modelling the entire TypeMapper around this, it may solve issues such as newer types introduced in later versions of a database.

I know that allowing the TypeMapper to query the database for types has been discussed before. I'm only raising it again because type aliases are bringing this back in a restricted way.

natemcmaster commented on src/EntityFramework.Relational.Design/IDatabaseModelFactory.cs:
```diff
@@ -8,6 +9,7 @@ namespace Microsoft.Data.Entity.Scaffolding
 {
     public interface IDatabaseModelFactory
     {
+        Dictionary<string, string> GetUserDefinedDataTypes([NotNull] string connectionString);
```
Since this is so SQL Server specific, we shouldn't put this on the interface.

lajones replied:
Fixed with 2nd commit.

natemcmaster commented on src/EntityFramework.MicrosoftSqlServer.Design/SqlServerDatabaseModelFactory.cs:
```diff
@@ -371,6 +371,47 @@ private void GetForeignKeys()
             }
         }
 
+        public virtual Dictionary<string, string> GetUserDefinedDataTypes(
```
This should just be part of `DatabaseModel` (See https://github.com/aspnet/EntityFramework/pull/3783).

lajones replied:
It is with the 2nd commit.

natemcmaster commented:
cc @ajcvickers (typemapping discussion)

natemcmaster commented on src/EntityFramework.MicrosoftSqlServer/Storage/Internal/SqlServerTypeMapper.cs:
```diff
@@ -110,12 +107,6 @@ public SqlServerTypeMapper()
 
         protected override string GetColumnType(IProperty property) => property.SqlServer().ColumnType;
 
-        protected override IReadOnlyDictionary<Type, RelationalTypeMapping> GetSimpleMappings()
```
This removal is causing build errors.

lajones replied:
Yep - now fixed.

lajones commented:
Suggest you all wait on the comments. As I mentioned above I'm going to re-do this based on the `TypeAlias` suggestion from @roji. I'm part way through and it addresses a lot of your comments @natemcmaster . Anyway I've removed the `needs-review tag` for now, I'll re-add it when I've updated the PR.

lajones commented:
@roji - this issue will just be for type aliases - as it is causing significant numbers of columns from SQL Server DBs to issue the "can't recognize the data type for this column so skipping it" warning message.

However all that does is say "this alias really means this underlying system data type, so map it to a C# type in the same way you would that type". If you'd like to have a general discussion about loading _all_ data types from the database itself I think that's a separate request - could you raise another issue if you'd like to discuss that please? At the moment I don't see how it would work - how would the DB know to which C# type to map its internal data types? But I could be missing something - so feel free to raise.

lajones commented:
@roji @natemcmaster Updated to a new commit using the `TypeAlias` approach and have re-added the `needs-review` tag.

lajones commented:
Chatted with @natemcmaster and decided we need a more SQL Server specific approach. Closing this. Will open another PR with that approach.

roji commented:
Dealing with this as an SQL Server-specific feature might be the way to go.

I'm personally interested in general in whether the type mapper should have access to a specific database, i.e. accepting a connection or connection string. Aside from SQL Server type aliases which seem to require this, PostgreSQL also allows types to be defined in a specific database etc.
